docs/developers/symbol-matcher.md

```diff
@@ -0,0 +1,179 @@
+---
+id: symbol-matcher
+title: SymbolMatcher
+---
+
+Symbol matcher is a trait that is defined by a single method
+`matches(Symbol): Boolean`.
+
+```scala
+trait SymbolMatcher {
+  def matches(symbol: Symbol): Boolean
+}
+```
+
+On top of `matches()`, symbol matchers provide additional methods to extract
+interesting tree nodes with pattern matching.
+
+All code examples in this document assume you have the following imports in
+scope.
+
+```scala mdoc
+import scalafix.v1._
+import scala.meta._
+```
+
+```scala mdoc:passthrough
+import scalafix.docs.PatchDocs
+import scalafix.docs.PatchDocs._
+```
+
+## Symbol specification
+
+To learn more about symbols, consult the SemanticDB specification:
+
+- [General symbols](https://github.com/scalameta/scalameta/blob/master/semanticdb/semanticdb3/semanticdb3.md#symbol)
+- [Scala symbols](https://github.com/scalameta/scalameta/blob/master/semanticdb/semanticdb3/semanticdb3.md#scala-symbol)
+- [Java symbols](https://github.com/scalameta/scalameta/blob/master/semanticdb/semanticdb3/semanticdb3.md#java-symbol)
+
+## SymbolMatcher reference
+
+There are two kinds of symbol matchers: exact and normalized. To illustrate the
+differences between exact and normalized symbol matchers, we use the following
+input source file as a running example.
+
+```scala mdoc:passthrough
+implicit var doc = PatchDocs.fromString("""
+package com
+class Main() {}            // com/Main# class
+object Main extends App {  // com/Main. object
+  println()                // println()   overload with no parameter
+  println("Hello Predef!") // println(+1) overload with string parameter
+}
+""")
+```
+
+Observe that the name
+
+- `Main` can refer to both the class or the companion object.
+- `println` can refer to different method overloads.
+
+### `exact()`
+
+Exact symbol matchers allow precise comparison between symbols with the ability
+to distinguish the differences between method overloads and a class from its
+companion object.
+
+To match only the `Main` class in the example above.
+
+```scala mdoc
+val mainClass  = SymbolMatcher.exact("com/Main#")
+mainClass.matches(Symbol("com/Main#"))
+mainClass.matches(Symbol("com/Main."))
+```
+
+To match only the `println(String)` method overload and not the `println()`
+method.
+
+```scala mdoc
+val printlnString  = SymbolMatcher.exact("scala/Predef.assert(+1).")
+printlnString.matches(Symbol("scala/Predef.assert()."))
+printlnString.matches(Symbol("scala/Predef.assert(+1)."))
```

Comment:
Did the example here change from assert to println?

Reply:
Good catch! Will fix in a followup PR, thanks ❤️